libsolidity/ast/TypeProvider.h

```diff
@@ -0,0 +1,224 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/Types.h>
+
+#include <array>
+#include <map>
+#include <memory>
+#include <utility>
+
+namespace dev
+{
+namespace solidity
+{
+
+/**
+ * API for accessing the Solidity Type System.
+ *
+ * This is the Solidity Compiler's type provider. Use it to request for types. The caller does
+ * <b>not</b> own the types.
+ *
+ * It is not recommended to explicitly instantiate types unless you really know what and why
+ * you are doing it.
+ */
+class TypeProvider
+{
+public:
+	TypeProvider();
+	TypeProvider(TypeProvider&&) = default;
+	TypeProvider(TypeProvider const&) = delete;
+	TypeProvider& operator=(TypeProvider&&) = default;
+	TypeProvider& operator=(TypeProvider const&) = delete;
+	~TypeProvider() = default;
+
+	/**
+	 * Resets state of this TypeProvider to initial state, wiping all mutable types.
+	 *
+	 * This invalidates all dangling pointers to types provided by this TypeProvider.
+	 */
+	void reset();
+
+	/**
+	 * Global TypeProvider instance.
+	 */
+	static TypeProvider& get()
+	{
+		static TypeProvider _provider;
+		return _provider;
+	}
+
```

Comment:
To argue my point: the next 150 lines or so are in fact duplicates of the ones in ``Types.h`` and that's only the declarations, not the definitions - adding or changing any constructor will now require at least four edits in two different files... moving all that to the types will probably not require less edits in the end, but they are at least in the same place. (One could use some template construction to mitigate all that either way, but still I think this stuff actually belongs in the types)